src/mlpack/methods/ann/ffn_impl.hpp

```diff
@@ -192,101 +122,254 @@ LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
       << "." << std::endl;
 }
 
-template<typename LayerTypes,
-         typename OutputLayerType,
-         typename InitializationRuleType,
-         typename PerformanceFunction
->
-void FFN<
-LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->::Predict(arma::mat& predictors, arma::mat& responses)
+template<typename OutputLayerType, typename InitializationRuleType>
+void FFN<OutputLayerType, InitializationRuleType>::Predict(
+    arma::mat& predictors, arma::mat& responses)
 {
-  deterministic = true;
+  if (parameter.is_empty())
+  {
+    ResetParameters();
+  }
+
+  if (!deterministic)
+  {
+    deterministic = true;
+    ResetDeterministic();
+  }
 
   arma::mat responsesTemp;
-  ResetParameter(network);
-  Forward(arma::mat(predictors.colptr(0), predictors.n_rows, 1, false, true),
-      network);
-  OutputPrediction(responsesTemp, network);
+  Forward(std::move(arma::mat(predictors.colptr(0),
+      predictors.n_rows, 1, false, true)));
+  responsesTemp = boost::apply_visitor(outputParameterVisitor,
+      network.back()).col(0);
 
   responses = arma::mat(responsesTemp.n_elem, predictors.n_cols);
   responses.col(0) = responsesTemp.col(0);
 
   for (size_t i = 1; i < predictors.n_cols; i++)
   {
-    Forward(arma::mat(predictors.colptr(i), predictors.n_rows, 1, false, true),
-        network);
+    Forward(std::move(arma::mat(predictors.colptr(i),
+        predictors.n_rows, 1, false, true)));
 
-    responsesTemp = arma::mat(responses.colptr(i), responses.n_rows, 1, false,
-        true);
-    OutputPrediction(responsesTemp, network);
+    responsesTemp = boost::apply_visitor(outputParameterVisitor,
+        network.back());
     responses.col(i) = responsesTemp.col(0);
   }
 }
 
-template<typename LayerTypes,
-         typename OutputLayerType,
-         typename InitializationRuleType,
-         typename PerformanceFunction
->
-double FFN<
-LayerTypes, OutputLayerType, InitializationRuleType, PerformanceFunction
->::Evaluate(const arma::mat& /* unused */,
-            const size_t i,
-            const bool deterministic)
+template<typename OutputLayerType, typename InitializationRuleType>
+double FFN<OutputLayerType, InitializationRuleType>::Evaluate(
+    const arma::mat& /* parameters */, const size_t i, const bool deterministic)
 {
-  this->deterministic = deterministic;
+  if (parameter.is_empty())
+  {
+    ResetParameters();
+  }
 
-  ResetParameter(network);
+  if (deterministic != this->deterministic)
+  {
+    this->deterministic = deterministic;
+    ResetDeterministic();
+  }
```

Comment:
The problem is, that some layer act differently in training and testing mode. Like the Dropout layer randomly sets some values to zero in training mode and uses all values while testing. Since the optimizer evaluates the network after each batch/iteration we have to set the deterministic parameter accordingly and sitch it back for the next iteration.

```
// Evaluate the gradient for this iteration.
if (shuffle)
  function.Gradient(iterate, visitationOrder[currentFunction], gradient);
else
  function.Gradient(iterate, currentFunction, gradient);
```

for the Gradient function, we have to set the deterministic parameter to training

```
// Now add that to the overall objective function.
if (shuffle)
  overallObjective += function.Evaluate(iterate,
      visitationOrder[currentFunction]);
else
  overallObjective += function.Evaluate(iterate, currentFunction);
```

for the Evaluate function, we have to set the deterministic parameter to testing.